internal/mode/static/nginx/config/upstreams_template.go

```diff
@@ -8,8 +8,12 @@ var upstreamsTemplateText = `
 upstream {{ $u.Name }} {
     random two least_conn;
     zone {{ $u.Name }} 512k;
-    {{ range $server := $u.Servers }}
+    {{- if $u.StateFile }}
+    state {{ $u.StateFile }};
```

Comment:
Can we potentially have race conditions?

(1)
For example, user deploys a Service and Endpoints

User adds an HTTPRoute that references that Service

(2)
User deploys an HTTPRoute and Service

User removes HTTPRoute

some time passes so that Service Endpoints change now

User adds HTTPRoute

Is there are chance NGINX Plus will use the old Endpoints here, not the latest ones?

Will NGINX Plus include all upstreams of that service?


Note: to ensure that happens, would it make sense to add e2e test for this feature?

Reply:
I see that now. thanks

I missed that after files are updated and NGINX (or NGINX Plus) is reloaded, we always update upstreams for NGINX Plus -- https://github.com/nginxinc/nginx-gateway-fabric/blob/cc4c98596ec720493228c0faee0e6a08240965ca/internal/mode/static/handler.go#L185-L206

Reply:
at the same time, for (2), I think there is a race.

between the reload of NGINX Plus and the update of the upstreams, NGINX Plus will use the last version of the state file. Which might contain IP addresses that no longer exist. Or perhaps even point to different backends .

Reply:
I don't think that's a race necessarily, more-so just an order of operations. Until an update of the upstreams is called, the old servers are going to be used. This is true whether or not a reload occurs.

Reply:
Let's imagine the state file was used 1 day ago. then it stop being used, because the HTTPRoute that referenced that service was removed.

during next 24 hours the Service is changes. For example, it is recreated, or perhaps some or all Endpoints of it changes.

potentially, Kubernetes also assigned IPs used by the terminated pods of that Service to new pods of some new Services.

now the user adds HTTPRoute that references that Service.

There will be a window after NGINX Plus reloaded before NGINX Plus upstreams are updated via an API, where NGINX Plus will serve the upstream servers from the previous version of the state file. As I mentioned, those IPs might not exist or could even be of some new pods of completely different Services.

Reply:
For what it's worth, NIC appears to do this the same way. I just go back to the inconsistency between nginx conf and actual state of the world if we don't use the state files. For support and debugging, it could make things hard to diagnose.

If we optimize the time between reload and API call, then we can limit the possibility for this race.

Reply:
I'd vote to eliminate the edge case by design even though it can hurt debugging.

Reply:
So essentially the suggestion is that if it's a ClusterChange (not an EndpointsOnly change), then we just reload and don't use the API at all. We keep the servers in the conf, and per the other conversation, if it's an EndpointsOnly change, we update the conf but don't reload, then use the API to update the upstreams.